[wip] add some opentelemetry support

akuzm opened the pull request:
some drafts for https://github.com/ClickHouse/ClickHouse/issues/13444

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added initial OpenTelemetry support. ClickHouse now accepts OpenTelemetry traceparent headers over Native and HTTP protocols, and passes them downstream in some cases. The trace spans for executed queries are saved into the `system.opentelemetry_span_log` table.

akuzm committed:
opentelemetry wip

akuzm committed:
fixup

akuzm committed:
remove accidental changes

akuzm committed:
opentelemetry context propagation

akuzm committed:
less crazy parsing

akuzm committed:
cosmetic fixes

akuzm committed:
A test + reinterpretAsUUID

akuzm committed:
remove some debug output to pass the tests

akuzm committed:
more tests + clickhouse-client options

akuzm committed:
fixes and some docs

akuzm committed:
Merge origin/master into tmp (using imerge)

Slach commented:
Hello @bobrik, you are work on another task ClickHouse as storage backend for jaeger
https://github.com/jaegertracing/jaeger/issues/1438
could you help us review these changes for add support ClickHouse OpenTelemetry client->Jaeger

bobrik commented on src/Interpreters/ClientInfo.cpp:
```diff
@@ -113,6 +132,26 @@ void ClientInfo::read(ReadBuffer & in, const UInt64 client_protocol_revision)
+            readBinary(opentelemetry_tracestate, in);
+            readBinary(opentelemetry_trace_flags, in);
+
+            fmt::print(stderr, "read {:x}, {}, {} at\n{}\n",
```
This looks like it shouldn't be here.

akuzm replied:
Yep, this draft still has some debug output here and there, will remove.

bobrik reviewed (commented):
This looks pretty exciting!

How do you envision integrating this with an existing system like Jaeger? Are end users expected to periodically read off tables, convert data into Jaeger format and ingest it? I haven't played with OpenTelemetry yet (we're still using plain old Jaeger internally), but have you considered using OTLP as a standard way to export tracing information?

* https://opentelemetry.io/docs/#exporting-data

nvartolomei commented on programs/server/config.xml:
```diff
@@ -563,6 +563,31 @@
+        <engine>
+            engine MergeTree
+            partition by toYYYYMM(finish_date)
+            order by (finish_date, finish_time, trace_id)
```
Can we remove finish_date? It makes queries a bit harder to type as usually you need to provide both date and time. Why not keep just finish_time?

akuzm replied:
A separate date column is a way to save I/O for filters like `event_date = today()` -- the time column can be 300 times bigger than date, and we can avoid reading it. We use it for all system logs. With OpenTelemetry log it may be less important -- you are probably going to store and explore it using your tracing system, not with ClickHouse SQL.

akuzm commented:
> Are end users expected to periodically read off tables, convert data into Jaeger format and ingest it?

Haven't thought this through yet.

There are basically two models, we can pull data from ClickHouse using some external app, or make ClickHouse itself push the data.

Pull is the simplest to implement in CH -- the log table is enough. But there has to be some service that polls CH and sends the data to metrics collector. This won't work out of the box, but is the most flexible. One thing that's not exactly clear here is how to determine which data we've sent already. Using `finish_time` as sequence number is mostly sufficient, but not always:

- it is system time, so may change non-monotonously
- even with monotonic time, there is nothing to guarantee that a trace span is not added into the past (e.g. we've flushed the `opentelemtry_log` up to 10:00:00, but on the next flush we write a span that finished at 09:59:59). Such races shouldn't occur often, but they are possible.

Internally, our system logs already use int64 global sequence numbers for messages, but this number is reset on restart, so it's also not sufficient by itself. 

Push is more complex. We do already push the metrics to Graphite (see [the `graphite` option in server settings](https://clickhouse.tech/docs/en/operations/server-configuration-parameters/settings/#server_configuration_parameters-graphite)). The sender of metrics is stateless, but the OT sender must also have a queue with persistence and backpressure and so on. The [OTLP spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/protocol/requirements.md) looks somewhat complicated, maybe we can implement a simpler protocol to send to a locally running third-party collector, which would then take care of the rest.

There is also a cheap, if somewhat weird, way to implement the push approach -- make a materialized view over `system.opentelemetry_log` that writes into a table with `URL` engine, that in turn sends the data to Jaeger, as JSON over HTTP ([Zipkin formats](https://www.jaegertracing.io/docs/1.19/apis/#zipkin-formats-stable) look suitable). I actually plan to try this for testing.

akuzm committed:
tests

akuzm committed:
Merge remote-tracking branch 'origin/master' into tmp

bobrik commented:
> even with monotonic time, there is nothing to guarantee that a trace span is not added into the past (e.g. we've flushed the `opentelemtry_log` up to 10:00:00, but on the next flush we write a span that finished at 09:59:59). Such races shouldn't occur often, but they are possible.

Sounds like we need a monotonic timestamp of the block itself becoming visible as a column to allow consumers to read in linear order, but I don't know if it's feasible.

For what it's worth, most of our traces come to a local agent over udp, which is pretty straighforward, but it's also quite opaque.

bobrik commented:
After looking around I found this:

* https://opentelemetry.io/docs/collector/configuration/

Jaeger collector also supports collecting spans in Zipkin format.

Let me know when you have a working Zipking-over-HTTP as a push target and I'll try to make my [Clickhouse backend for Jaeger](https://github.com/jaegertracing/jaeger/issues/1438) integrate with this, so we have traces with spans from both Jaeger UI and its Clickhouse backend.

bobrik commented on src/Interpreters/OpenTelemetryLog.cpp:
```diff
@@ -0,0 +1,46 @@
+        {std::make_shared<DataTypeUInt64>(), "span_id"},
+        {std::make_shared<DataTypeUInt64>(), "parent_span_id"},
+        {std::make_shared<DataTypeString>(), "operation_name"},
+        {std::make_shared<DataTypeDateTime>(), "start_time"},
```
Timestamps should get at least microsecond precision, at second level granularity it's much harder to figure out what is going on in terms of order of events.

akuzm replied:
Extended them to microsecond precision.

bobrik commented on src/Interpreters/OpenTelemetryLog.cpp:
```diff
@@ -0,0 +1,46 @@
+Block OpenTelemetrySpanLogElement::createBlock()
+{
+    return {
+        {std::make_shared<DataTypeUUID>(), "trace_id"},
```
Trace ID seems to be a 128 bit integer (32 char hex [according to spec](https://www.w3.org/TR/trace-context/#versioning-of-traceparent)).

UUIDv4 is a 128 bit integer and it can be converted to a hex via `hex(UUIDStringToNum(toString(trace_id)))`, but I'm not sure if it's the right logical format to store the value here.

akuzm replied:
We got big int types recently, but they are still very experimental, so I'll keep the UUID for now.

bobrik commented on src/Interpreters/Context.cpp:
```diff
@@ -1083,28 +1083,53 @@ void Context::setCurrentQueryId(const String & query_id)
+    {
+        // The incoming request has an OpenTelemtry trace context. Its span id
+        // becomes our parent span id.
+        client_info.opentelemetry_parent_span_id = client_info.opentelemetry_span_id;
```
Not sure if this is right. When I send a query:

```
echo 'select 1' | clickhouse-client --opentelemetry-traceparent=00-000000000000000057b17940b00d1173-0000000000000073-03
```

In the table it has `parent_span_id` set to zero, when it should be `0x73`.

This is the place where we set `parent_span_id` for queries with an incoming context, and `span_id` from client info is a value initialized to zero and only set to a non-zero value in the `if` branch above.

My guess is that we should just remove this line.

akuzm replied:
I fixed this case but in a different way -- by setting span_id instead of parent_span_id from traceparent header. This way, ser/de is symmetric and consistent with native protocol -- we always serialize span_id.

akuzm committed:
Merge remote-tracking branch 'origin/master' into tmp

akuzm committed:
fix the uuid test

akuzm committed:
cleanup

akuzm committed:
straighten the protocol version

akuzm committed:
Merge remote-tracking branch 'origin/aku/tcp-versio' into tmp

akuzm committed:
Merge remote-tracking branch 'origin/master' into tmp

akuzm committed:
Merge remote-tracking branch 'origin/master' into tmp

akuzm committed:
Merge remote-tracking branch 'origin/master' into tmp

akuzm committed:
allow fast test to run locally

akuzm committed:
fasttest fixup

akuzm committed:
fasttest fixup 2

akuzm committed:
fasttest fixup 3

akuzm committed:
fasttest fixup 3

akuzm committed:
fasttest compat with old script

akuzm committed:
microsecond precision for start/finish time

akuzm committed:
fasttest fixup 5

akuzm committed:
don't shadow local var

akuzm committed:
Merge remote-tracking branch 'origin/master' into HEAD

akuzm committed:
fixup

akuzm committed:
Merge remote-tracking branch 'origin/master' into HEAD

akuzm committed:
Merge remote-tracking branch 'origin/aku/error-context' into HEAD

akuzm committed:
Merge origin/master into tmp (using imerge)

akuzm committed:
more fasttest fixes

akuzm committed:
fixup

akuzm committed:
Merge remote-tracking branch 'origin/master' into tmp

akuzm committed:
Merge remote-tracking branch 'origin/master' into tmp

akuzm committed:
fix the test

akuzm committed:
improve fasttest usability

akuzm committed:
reset error code in ast fuzzer

akuzm committed:
remove warnings in integration tests

akuzm committed:
Merge remote-tracking branch 'origin/master' into tmp

akuzm committed:
fix initial query id

akuzm committed:
disable traceparent header in Arcadia

it interferes with the test_clickhouse.TestTracing.test_tracing_via_http_proxy[traceparent] test

akuzm committed:
fix setting parent from HTTP traceparent

akuzm committed:
cleanup

akuzm committed:
Merge remote-tracking branch 'origin/master' into tmp

akuzm commented:
I intend to merge this soon. The main added value is that ClickHouse now accepts the traceparent header over Native and HTTP protocols, and passes it downstream in some cases.

Tracing of the ClickHouse internals is incipient, and integration with trace collectors is nonexistent.

akuzm committed:
try to fix weird docs error

akuzm committed:
Merge remote-tracking branch 'origin/master' into HEAD

akuzm commented:
testflows -- spurios failure https://github.com/ClickHouse/ClickHouse/pull/16427
